src/diagnostics.md

```diff
@@ -61,6 +61,163 @@ error: the fobrulator needs to be krontrificated
 When code or an identifier must appear in an message or label, it should be
 surrounded with single acute accents \`.
 
+### Error explanations
+
+Some errors include long form descriptions. They may be viewed with the
+`--explain` flag, or via the [error index]. Each explanation comes with an
+example of how to trigger it and advice on how to fix it.
+
+Please read [RFC 1567] for details on how to format and write long error
+codes.
+
+The descriptions are written in markdown, and all of them are linked in the
+[`librustc_error_codes`] crate.
+
+<!-- TODO: When should an error use an error code, and when shouldn't it? -->
+
+[`librustc_error_codes`]: https://github.com/rust-lang/rust/blob/master/src/librustc_error_codes/error_codes.rs
+[error index]: https://doc.rust-lang.org/error-index.html
+[RFC 1567]: https://github.com/rust-lang/rfcs/blob/master/text/1567-long-error-codes-explanation-normalization.md
+
+### Lints versus fixed diagnostics
+
+Some messages are emitted via [lints](#lints), where the user can control the
+level. Some are hard-coded such that the user cannot control the level.
```

Comment:
Right, I'm not sure how to phrase it (is there existing terminology?).  Just comparing when to create a lint versus when to make a fixed diagnostic with the `span_*` methods.  Usually it is obvious (invalid code is always a fixed error), but there are some grey areas.

Reply:
Ah, I see what you mean now. Perhaps it would be helpful to give a few examples at the end? For example, something like the following:

> Usually it is obvious whether a diagnostic should be "fixed" or a lint, but there are some grey areas.
> 
> Here are a few examples:
>
> - Borrow checker errors: these are fixed errors. The user cannot adjust the level of these diagnostics to silence the borrow checker.
> - Dead code: this is a lint. While the user probably doesn't want dead code in their crate, making this a hard error would make refactoring and development very painful.
> - [`safe_packed_borrows` future compatibility warning](https://github.com/rust-lang/rust/issues/46043): this is a silencable lint related to safety. It was judged that the making this a hard (fixed) error would cause too much breakage, so instead a warning is emitted that eventually will be turned into a hard error.

Reply:
Thanks, added the suggestion, it looks good!